src/shared/utilities/childProcess.ts

```diff
@@ -45,138 +76,153 @@ export class ChildProcess {
+            try {
+                this.childProcess = crossSpawn.spawn(this.command, args, mergedOptions.spawnOptions)
+            } catch (err) {
+                return errorHandler(err as Error)
             }
-        })
+
+            // Emitted whenever:
+            //  1. Process could not be spawned, or
+            //  2. Process could not be killed, or
+            //  3. Sending a message to the child process failed.
+            // https://nodejs.org/api/child_process.html#child_process_class_childprocess
+            // We also register error event handlers on the output/error streams in case a lower level library fails
+            this.childProcess.on('error', errorHandler)
+            this.childProcess.stdout?.on('error', errorHandler)
+            this.childProcess.stderr?.on('error', errorHandler)
+
+            this.childProcess.stdout?.on('data', (data: { toString(): string }) => {
+                if (mergedOptions.collect) {
+                    this.stdoutChunks.push(data.toString())
+                }
+
+                mergedOptions.onStdout?.call(paramsContext, data.toString())
+            })
+
+            this.childProcess.stderr?.on('data', (data: { toString(): string }) => {
+                if (mergedOptions.collect) {
+                    this.stderrChunks.push(data.toString())
+                }
+
+                mergedOptions.onStderr?.call(paramsContext, data.toString())
+            })
+
+            // Emitted when streams are closed.
+            this.childProcess.once('close', (code, signal) => {
+                this.processResult = this.makeResult(code, signal)
+                resolve(this.processResult)
+                cleanup()
+            })
+
+            // Emitted when process exits or terminates.
+            // https://nodejs.org/api/child_process.html#child_process_class_childprocess
+            // - If the process exited, `code` is the final exit code of the process, else null.
+            // - If the process terminated because of a signal, `signal` is the name of the signal, else null.
+            // - One of `code` or `signal` will always be non-null.
+            this.childProcess.once('exit', (code, signal) => {
+                this.processResult = this.makeResult(
+                    typeof code === 'number' ? code : -1,
+                    typeof signal === 'string' ? signal : undefined
+                )
+                if (code && rejectOnExit) {
+                    if (typeof rejectOnExit === 'function') {
+                        reject(rejectOnExit(code))
+                    } else {
+                        reject(new Error(`Command exited with non-zero code: ${code}`))
+                    }
+                }
+                if (mergedOptions.waitForStreams === false) {
+                    resolve(this.processResult)
+                }
+            })
+        }).finally(() => this.childProcess?.killed && cleanup())
```

Comment:
I'm not sure why `this.childProcess?.killed &&` is here? `finally` won't return anything, and we're not in charge of setting the `killed` value.

Reply:
We want to remove listeners on Promise fulfillment since there really isn't a reason to keep them. The stdout/stderr can be shared by multiple processes so it's a good check to have. Now that I think about it, we should drop the `result` method from the class and favor an immutable result instead.

The syntax here is just using expressions to conditionally evaluate `cleanup` (like a shell script)

Reply:
Huh, didn't realize this syntax worked in JS. What happens if the process isn't in a `killed` state when this is hit, will it orphan the listeners?

Reply:
Yeah it would. I think I had the `killed` check here because I thought some code was using this behavior in some way, though that doesn't seem to be the case. I can't really think of a use-case where you'd want the promise to resolve/reject but still let the child process to continue? 

For the SAM debug stuff I know we check for a 'ready to attach' message but that can be achieved with a separate promise.

Reply:
If it's that important to clean up the listeners, I'd rather always call `cleanup` and just kill them if they're defined there.

Reply:
Not properly disposing of things doesn't cause issues the vast majority of time, but when it does it's very obnoxious to debug. So yeah I'll say if the promise resolve/rejects the listeners are done.